Add PropertyWrapperFontWeight to bound the values for weight

chirags27 opened the pull request:
#### 3070b3d704a3a9364da1442c873419a7b6757343
<pre>
Add PropertyWrapperFontWeight to bound the values for weight
<a href="https://bugs.webkit.org/show_bug.cgi?id=247140">https://bugs.webkit.org/show_bug.cgi?id=247140</a>
rdar://problem/100986499

Reviewed by Myles C. Maxfield.

Currently, fontWeight wasn&apos;t bounded based on the CSS specs, which led
to scenarios where FontSelectionRequest had a 0 fontWeight, and which
resulted in an empty FontSelectionRequest, which is an invalid key
based for the HashMap segmentedFontFaceCache. This resulted in map
corruption, which led to a crash. The change fixes the issue by
correctly defining PropertyWrapperFontWeight, which doesn&apos;t allow the
blend() to set invalid weight values.

Test: fonts/font-weight-invalid-crash.html

* LayoutTests/fonts/font-weight-invalid-crash-expected.txt: Added.
* LayoutTests/fonts/font-weight-invalid-crash.html: Added.
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::blendFunc):
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/animations/system-fonts-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-fonts/animations/system-fonts-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-fonts/animations/system-fonts-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-fonts/animations/system-fonts-expected.txt:

Canonical link: <a href="https://commits.webkit.org/256123@main">https://commits.webkit.org/256123@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f0f518d361094f92a82661d9b1872739140281f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94767 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3921 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/27642 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/104376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/164640 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98763 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3993 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/32102 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/87073 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/100300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100434 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/2876 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/81122 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/87073 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/84804 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/84346 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/87073 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38512 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18160 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/36349 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/19440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4223 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40270 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42087 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42244 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/38671 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

litherum commented on LayoutTests/fonts/font-weight-invalid-crash.html:
```diff
@@ -0,0 +1,32 @@
+*:root,.class0 { separate;font-stretch: expanded;-webkit-padding-before: 'a b';-webkit-column-break-inside: farthest-side invert auto;font-weight: 1;inset-inline-end: }
+progress,*:scope,.class3 { 0px;animation: keyframes4,keyframes1 cubic-bezier(0,358,1,1) }
+#x50:placeholder-shown,#x73:dir(ltr),*:defined { space;-webkit-animation-duration: 1s;border-bottom-style: auto;transition-duration: 1.5s;rotation: 180deg;-webkit-transition-timing-function: step-start,cubic-bezier(1,0,0,16);right: }
+*::before { -webkit-text-orientation: upright;-webkit-transition-property: }
```
We shouldn’t need all of this stuff just to test this one change. I recommend creating a new test from scratch.

chirags27 replied:
I've reduced the test as much possible, and further reducing it doesn't cause the issue to reproduce.

graouts commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -2342,6 +2341,22 @@ class PropertyWrapperSVGPaint final : public AnimationPropertyWrapperBase {
+
+    void blend(RenderStyle& destination, const RenderStyle& from, const RenderStyle& to, const CSSPropertyBlendingContext& context) const final
+    {
+        (destination.*m_setter)(FontSelectionValue(std::min(std::max(1.0f, blendFunc(static_cast<float>(this->value(from)), static_cast<float>(this->value(to)), context)), 1000.0f)));
```
What is the `1000` value coming from? Might it be more appropriate to be a named constant?

chirags27 replied:
It is based on the spec here: https://w3c.github.io/csswg-drafts/css-fonts-4/#valdef-font-weight-number-1-1000

I'll update it to using named constants!

graouts commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -3280,7 +3295,7 @@ CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap()
 #endif
         new PropertyWrapperFontSizeAdjust,
-        new PropertyWrapper<FontSelectionValue>(CSSPropertyFontWeight, &RenderStyle::fontWeight, &RenderStyle::setFontWeight),
+        new PropertyWrapperFontWeight(),
```
No need for the `()` since we're not passing any argument.

```suggestion
        new PropertyWrapperFontWeight
```

graouts commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -2342,6 +2341,22 @@ class PropertyWrapperSVGPaint final : public AnimationPropertyWrapperBase {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    PropertyWrapperFontWeight()
+        : PropertyWrapper<FontSelectionValue>(CSSPropertyFontWeight, &RenderStyle::fontWeight, &RenderStyle::setFontWeight)
```
I don't think you need the `<FontSelectionValue>` here and that this would be sufficient:

```suggestion
        : PropertyWrapper(CSSPropertyFontWeight, &RenderStyle::fontWeight, &RenderStyle::setFontWeight)
```

chirags27 committed:
Add PropertyWrapperFontWeight to bound the values for weight
https://bugs.webkit.org/show_bug.cgi?id=247140
rdar://problem/100986499

Reviewed by Myles C. Maxfield.

Currently, fontWeight wasn't bounded based on the CSS specs, which led
to scenarios where FontSelectionRequest had a 0 fontWeight, and which
resulted in an empty FontSelectionRequest, which is an invalid key
based for the HashMap segmentedFontFaceCache. This resulted in map
corruption, which led to a crash. The change fixes the issue by
correctly defining PropertyWrapperFontWeight, which doesn't allow the
blend() to set invalid weight values.

Test: fonts/font-weight-invalid-crash.html

* LayoutTests/fonts/font-weight-invalid-crash-expected.txt: Added.
* LayoutTests/fonts/font-weight-invalid-crash.html: Added.
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::blendFunc):
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/animations/system-fonts-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-fonts/animations/system-fonts-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-fonts/animations/system-fonts-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-fonts/animations/system-fonts-expected.txt:

Canonical link: https://commits.webkit.org/256123@main
